mesonbuild/scripts/reprotest.py

```diff
@@ -0,0 +1,130 @@
+                           '.obj',
+                           '.class',
+                           )
+
+class ReproTester:
+    def __init__(self, options: T.Any):
+        self.args = options.mesonargs
+        self.meson = get_meson_command()[:]
+        self.builddir = pathlib.Path('buildrepro')
+        self.storagedir = pathlib.Path('buildrepro.1st')
+        self.issues: T.List[str] = []
+        self.check_intermediaries = options.intermediaries
+
+    def run(self) -> int:
+        if not pathlib.Path('meson.build').is_file():
+            sys.exit('This command needs to be run at your project source root.')
+        self.check_ccache()
+        self.cleanup()
+        self.build()
+        self.check_output()
+        self.print_results()
+        if not self.issues:
+            self.cleanup()
+        return len(self.issues)
+
+    def check_ccache(self) -> None:
+        for evar in ('CC', 'CXX'):
+            evalue = os.environ.get(evar, '')
+            if 'ccache' in evalue:
+                print(f'Environment variable {evar} set to value "{evalue}".')
+                print('This implies using a compiler cache, which is incompatible with reproducible builds.')
+                sys.exit(1)
+
+    def cleanup(self) -> None:
+        if self.builddir.exists():
+            shutil.rmtree(self.builddir)
+        if self.storagedir.exists():
+            shutil.rmtree(self.storagedir)
+
+    def build(self) -> None:
+        setup_command: T.Sequence[T.Union[pathlib.Path, str]] = self.meson + ['setup', self.builddir] + self.args
+        build_command: T.Sequence[T.Union[pathlib.Path, str]] = self.meson + ['compile', '-C', self.builddir]
+        subprocess.check_call(setup_command)
+        subprocess.check_call(build_command)
+        self.builddir.rename(self.storagedir)
+        subprocess.check_call(setup_command)
+        subprocess.check_call(build_command)
+
+    def ignore_file(self, fstr: str) -> bool:
+        for p in IGNORE_PATTERNS:
+            if p in fstr:
+                return True
+        if not self.check_intermediaries:
+            for e in INTERMEDIATE_EXTENSIONS:
+                if fstr.endswith(e):
+                    return True
+        return False
+
+    def check_contents(self, fromdir: str, todir: str, check_contents: bool) -> None:
+        frompath = fromdir + '/'
+        topath = todir + '/'
+        for fromfile in pathlib.Path(fromdir).glob('**/*'):
+            if not fromfile.is_file():
+                continue
+            fstr = str(fromfile)
+            if self.ignore_file(fstr):
+                continue
+            assert fstr.startswith(frompath)
+            tofile = pathlib.Path(fstr.replace(frompath, topath, 1))
+            if not tofile.exists():
+                self.issues.append(f'Missing file: {tofile}')
+            elif check_contents:
+                fromdata = fromfile.read_bytes()
+                todata = tofile.read_bytes()
+                if fromdata != todata:
```

Comment:
Please use the filecmp module for this, it is designed to compare two files to see if they are identical while short circuiting as soon as it detects a difference.

In the common case it short circuits without ever opening the file for reading because the filesystem stat result says the files are different sizes and this cannot be identical. Otherwise it performs chunked reads rather than reading the whole file into memory and returns early as soon as the first chunk differs.

Reply:
There is also a directory comparison API which may be able to replace the glob iteration.

Reply:
> There is also a directory comparison API which may be able to replace the glob iteration.

Sadly no. It requires a pregenerated list of file names to check. In order to build that you have to traverse both directories already. Given this it is simpler to just print the mismatches immediately.